services/src/main/java/org/keycloak/userprofile/DeclarativeUserProfileProvider.java

```diff
@@ -292,20 +292,19 @@ protected UserProfileMetadata decorateUserProfileForCache(UserProfileMetadata de
             }
 
             UPAttributeRequired rc = attrConfig.getRequired();
-            Predicate<AttributeContext> required = AttributeMetadata.ALWAYS_FALSE;
+            if (rc != null) {
+                validators.add(new AttributeValidatorMetadata(AttributeRequiredByMetadataValidator.ID));
+            }
 
+            Predicate<AttributeContext> required = AttributeMetadata.ALWAYS_FALSE;
             if (rc != null && !isUsernameOrEmailAttribute(attributeName)) {
-                // do not take requirements from config for username and email as they are
-                // driven by business logic from parent!
```

Comment:
@pedroigor I think this comment is outdated, thus I removed it. Since you added it originally, please double check before merging.